contracts/contracts/token/OUSD.sol

```diff
@@ -445,159 +441,268 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
     }
 
     /**
-     * @notice Enable rebasing for an account.
-     * @dev Add a contract address to the non-rebasing exception list. The
-     * address's balance will be part of rebases and the account will be exposed
-     * to upside and downside.
+     * @notice The calling account will start receiving yield after a successful call.
      * @param _account Address of the account.
      */
-    function governanceRebaseOptIn(address _account)
-        public
-        nonReentrant
-        onlyGovernor
-    {
+    function governanceRebaseOptIn(address _account) external onlyGovernor {
+        require(_account != address(0), "Zero address not allowed");
         _rebaseOptIn(_account);
     }
 
     /**
-     * @dev Add a contract address to the non-rebasing exception list. The
-     * address's balance will be part of rebases and the account will be exposed
-     * to upside and downside.
+     * @notice The calling account will start receiving yield after a successful call.
      */
-    function rebaseOptIn() public nonReentrant {
+    function rebaseOptIn() external {
         _rebaseOptIn(msg.sender);
     }
 
     function _rebaseOptIn(address _account) internal {
-        require(_isNonRebasingAccount(_account), "Account has not opted out");
-
-        // Convert balance into the same amount at the current exchange rate
-        uint256 newCreditBalance = _creditBalances[_account]
-            .mul(_rebasingCreditsPerToken)
-            .div(_creditsPerToken(_account));
+        uint256 balance = balanceOf(_account);
 
-        // Decreasing non rebasing supply
-        nonRebasingSupply = nonRebasingSupply.sub(balanceOf(_account));
-
-        _creditBalances[_account] = newCreditBalance;
-
-        // Increase rebasing credits, totalSupply remains unchanged so no
-        // adjustment necessary
-        _rebasingCredits = _rebasingCredits.add(_creditBalances[_account]);
+        // prettier-ignore
+        require(
+            alternativeCreditsPerToken[_account] > 0 ||
+                // Accounts may explicitly `rebaseOptIn` regardless of
+                // accounting if they have a 0 balance.
+                balance == 0
+            ,
+            "Account must be non-rebasing"
+        );
+        RebaseOptions state = rebaseState[_account];
+        // prettier-ignore
+        require(
+            state == RebaseOptions.StdNonRebasing ||
+                state == RebaseOptions.NotSet,
+            "Only standard non-rebasing accounts can opt in"
+        );
 
-        rebaseState[_account] = RebaseOptions.OptIn;
+        // Account
+        rebaseState[_account] = RebaseOptions.StdRebasing;
+        alternativeCreditsPerToken[_account] = 0;
+        (uint256 newCredits, uint256 newBalance) = _balanceToRebasingCredits(
+            balance
+        );
+        creditBalances[_account] = newCredits;
+        // Globals
+        _adjustGlobals(newCredits.toInt256(), -newBalance.toInt256());
```

Comment:
🔴 This results in incorrect accounting once we hit low resolution.

`_adjustGlobals` needs to work off the before and after differences in these global values.

RebaseOptIn moves an account from nonRebasing to rebasing. The global `rebasingCredits` will go up by amount credits the account gets, while the global `nonRebasingSupply` will go down by the **old** balance that the account used to hold. This current code adjusts by the new balance, but the new balance is not what is being changed.

Simplified test case"

```solidity
// low res token
ousd.initialize(address(this), 1e14-1); 
 
vm.prank(nonrebasing);
ousd.rebaseOptOut();

// Has to happen between the optOut and the OptIn
ousd.changeSupply(ousd.totalSupply()+1e19+1e11+2);

// Reverts because the math is not good.
vm.prank(nonrebasing);
ousd.rebaseOptIn();
```

Reply:
Corrected this, and other problems around globals in https://github.com/OriginProtocol/origin-dollar/pull/2298/commits/5e57112a9db3bed51d5ef725a589ca89f576df45

Reply:
This is a very cool find. So the `rebaseOptIn` failed with an underflow because the `rebasingCredits` are rounded up and with low resolution that would cause the newly computed balance in `_balanceToRebasingCredits` to be larger than the global `nonRebasingSupply `.